pennylane/_qubit_device.py

```diff
@@ -1670,35 +1609,29 @@ def marginal_prob(self, prob, wires=None):
         device_wires = self.map_wires(wires)
         inactive_device_wires = self.map_wires(inactive_wires)
 
-        # reshape the probability so that each axis corresponds to a wire
-        shape = [2] * self.num_wires
-        if batch_size is not None:
-            shape.insert(0, batch_size)
-        # prob now is reshaped to have self.num_wires+1 axes in the case of broadcasting
-        prob = self._reshape(prob, shape)
-
-        # sum over all inactive wires
         # hotfix to catch when default.qubit uses this method
         # since then device_wires is a list
         if isinstance(inactive_device_wires, Wires):
             inactive_device_wires = inactive_device_wires.labels
 
+        # reshape the probability so that each axis corresponds to a wire
+        shape = [2] * self.num_wires
+        desired_axes = np.argsort(np.argsort(device_wires))
+        flat_shape = (-1,)
         if batch_size is not None:
+            shape.insert(0, batch_size)
             inactive_device_wires = [idx + 1 for idx in inactive_device_wires]
-        flat_shape = (-1,) if batch_size is None else (batch_size, -1)
-        prob = self._reshape(self._reduce_sum(prob, inactive_device_wires), flat_shape)
-
-        # The wires provided might not be in consecutive order (i.e., wires might be [2, 0]).
-        # If this is the case, we must permute the marginalized probability so that
-        # it corresponds to the orders of the wires passed.
-        num_wires = len(device_wires)
-        basis_states = self.generate_basis_states(num_wires)
-        basis_states = basis_states[:, np.argsort(np.argsort(device_wires))]
+            desired_axes = np.insert(desired_axes + 1, 0, 0)
+            flat_shape = (batch_size, -1)
 
-        powers_of_two = 2 ** np.arange(len(device_wires))[::-1]
-        perm = basis_states @ powers_of_two
-        # The permutation happens on the last axis both with and without broadcasting
-        return self._gather(prob, perm, axis=1 if batch_size is not None else 0)
+        # prob now is reshaped to have self.num_wires+1 axes in the case of broadcasting
+        prob = self._reshape(prob, shape)
+        # sum over all inactive wires
+        prob = self._reduce_sum(prob, inactive_device_wires)
+        # rearrange wires to desired order
+        prob = self._transpose(prob, desired_axes)
```

Comment:
Good finding, how efficient is this?

Reply:
just tried something basic, it seems to be notably slower than on master... but master is wrong :) I can try something new like what was being done by the Cirq plugin (linked above). FWIW, here's my basic profiling technique:
```python
state = np.array([1 / 2, 0] * 4)
dev_wires = qml.wires.Wires(range(3))

def profile_wires(state, dev_wires):
    for probs_wires in ((0,1), (1, 2), (0, 2), (2, 0), (2, 1), (1, 2, 0), (2, 1, 0)):
        _ = qml.probs(wires=probs_wires).process_state(state, dev_wires)

%timeit profile_wires(state, dev_wires)
```

Master: `416 µs ± 5.22 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)`
Branch: `607 µs ± 9.81 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)`

Reply:
Feel free to explore :+1:  I do not think it is that bad, because prob should be called only once and now we have correct results.

Reply:
update! it wasn't transpose causing this difference, but instead the `if not np.all(desired_axes == range(len(desired_axes))):` check (I only did this check in `probs.py`, not for default.qubit). I removed the check in the latest commit, now the branch is more like 450us which is barely an overhead.

Reply:
Excellent @timmysilv :100: